crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -1320,6 +1330,77 @@ impl<'db> TypeInferenceBuilder<'db> {
         );
     }
 
+    fn infer_typevar_definition(
+        &mut self,
+        node: &ast::TypeParamTypeVar,
+        definition: Definition<'db>,
+    ) {
+        let ast::TypeParamTypeVar {
+            range: _,
+            name,
+            bound,
+            default,
+        } = node;
+        let bound_or_constraint = match bound.as_deref() {
+            Some(expr @ ast::Expr::Tuple(ast::ExprTuple { elts, .. })) => {
+                let tys = elts
+                    .iter()
+                    .map(|expr| self.infer_type_expression(expr))
+                    .collect::<Vec<_>>();
+                let constraints =
+                    TypeVarBoundOrConstraints::Constraints(tys.clone().into_boxed_slice());
+                self.store_expression_type(
+                    expr,
+                    Type::Tuple(TupleType::new(self.db, tys.into_boxed_slice())),
+                );
+                Some(constraints)
+            }
+            Some(expr) => Some(TypeVarBoundOrConstraints::UpperBound(
+                self.infer_type_expression(expr),
+            )),
+            None => None,
+        };
+        let default_ty = self.infer_optional_type_expression(default.as_deref());
+        self.add_binding(
+            node.into(),
+            definition,
+            Type::KnownInstance(KnownInstanceType::TypeVar(TypeVarInstance::new(
+                self.db,
+                name.id.clone(),
```

Comment:
`Name` implements `salsa::interned::Lookup`. You can see how it is used in the `FunctionType::new` calls

Reply:
It looks like if I turn `TypeVarInstance` into a tracked struct instead of an interned one, this doesn't work anymore.